src/sysupdate/updatectl.c

```diff
@@ -0,0 +1,1257 @@
+        _cleanup_(sd_event_unrefp) sd_event *event = NULL;
+        _cleanup_strv_free_ char **targets = NULL, **target_paths = NULL;
+        size_t n;
+        unsigned remaining = 0;
+        int r;
+
+        r = ensure_targets(bus, argv + 1, &targets);
+        if (r < 0)
+                return log_error_errno(r, "Failed to find targets: %m");
+
+        r = parse_targets(targets, &n, &target_paths, /* ret_versions= */ NULL);
+        if (r < 0)
+                return log_error_errno(r, "Failed to parse targets: %m");
+
+        table = table_new("target", "update");
+        if (!table)
+                return log_oom();
+
+        (void) table_set_sort(table, 0);
+
+        r = sd_event_default(&event);
+        if (r < 0)
+                return log_error_errno(r, "Failed to get event loop: %m");
+
+        r = sd_bus_attach_event(bus, event, 0);
+        if (r < 0)
+                return log_error_errno(r, "Failed to attach bus to event loop: %m");
+
+        r = sd_event_set_signal_exit(event, true);
+        if (r < 0)
+                return log_error_errno(r, "Failed to set up interrupt handler: %m");
+
+        for (size_t i = 0; i < n; i++) {
+                _cleanup_(operation_freep) Operation *op = NULL;
+                op = operation_new(table, bus, &remaining, target_paths[i], targets[i]);
+                if (!op)
+                        return log_oom();
+
+                r = sd_bus_call_method_async(bus, NULL, bus_sysupdate_mgr->destination, target_paths[i], SYSUPDATE_TARGET_INTERFACE, "CheckNew", check_finished, op, NULL);
+                if (r < 0)
+                        return log_error_errno(r, "Failed to call CheckNew for target %s: %m", targets[i]);
+                TAKE_PTR(op);
+
+                remaining++;
+        }
+
+        r = sd_event_loop(event);
+        if (r < 0)
+                return log_error_errno(r, "Failed to start event loop: %m");
+
+        return table_print_with_pager(table, SD_JSON_FORMAT_OFF, arg_pager_flags, arg_legend);
+}
+
+#define UPDATE_PROGRESS_FAILED INT_MIN
+/* Make sure it doesn't overlap w/ errno values */
+assert_cc(UPDATE_PROGRESS_FAILED < -ERRNO_MAX);
+
+static int update_render_progress(sd_event_source *source, void *userdata) {
+        OrderedHashmap *map = ASSERT_PTR(userdata);
+        const char *target;
+        void *p;
+        unsigned total;
+        size_t n;
+        bool exiting;
+
+        exiting = sd_event_get_state(sd_event_source_get_event(source)) == SD_EVENT_EXITING;
+
+        total = 0;
+        n = ordered_hashmap_size(map);
+
+        if (n == 0)
+                return 0;
+
+        if (!terminal_is_dumb()) {
+                for (size_t i = 0; i <= n; i++) {
+                        fputs("\n", stderr); /* Possibly scroll the terminal to make room (including total)*/
+                }
+                fprintf(stderr, "\x1B[%zuF", n+1); /* Go back */
+
+                fputs("\x1B""7", stderr); /* Save cursor position */
```

Comment:
the inner "" is redudant

Reply:
Without the inner "" compilers interpret this as a hex escape of `0x1B7` and complain it's out of range -- GCC gives a warning and clang an error.

```
$ cat test.c
#include "stdio.h"

int main(void) {
        fputs("\x1B7", stderr);
        return 0;
}
$ gcc test.c
test.c: In function ‘main’:
test.c:4:22: warning: hex escape sequence out of range
    4 |         fputs("\x1B7", stderr);
      |                      ^
$ clang test.c
test.c:4:16: error: hex escape sequence out of range
    4 |         fputs("\x1B7", stderr);
      |                ^~~~~
1 error generated.
```

Reply:
Done, I never heard of `\e` escapes before!